integration/webhook_test.go

```diff
@@ -0,0 +1,163 @@
+				Type: addonsv1alpha1.OLMOwnNamespace,
+			}, addonName),
+			err: testutil.NewStatusError(".spec.install.ownNamespace is required " +
+				"when .spec.install.type = OwnNamespace"),
+		},
+		{
+			addon: testutil.NewAddonWithInstallSpec(addonsv1alpha1.AddonInstallSpec{
+				Type: addonsv1alpha1.OLMAllNamespaces,
+			}, addonName),
+			err: testutil.NewStatusError(".spec.install.allNamespaces is required " +
+				"when .spec.install.type = AllNamespaces"),
+		},
+		{
+			addon: testutil.NewAddonWithInstallSpec(addonsv1alpha1.AddonInstallSpec{
+				Type: addonsv1alpha1.OLMOwnNamespace,
+				OLMOwnNamespace: &addonsv1alpha1.AddonInstallOLMOwnNamespace{
+					AddonInstallOLMCommon: addonsv1alpha1.AddonInstallOLMCommon{
+						Namespace:          "reference-addon",
+						PackageName:        addonName,
+						Channel:            "alpha",
+						CatalogSourceImage: referenceAddonCatalogSourceImageWorking,
+					},
+				},
+			}, addonName),
+			err: nil,
+		},
+		{
+			addon: testutil.NewAddonWithInstallSpec(addonsv1alpha1.AddonInstallSpec{
+				Type: addonsv1alpha1.OLMAllNamespaces,
+				OLMAllNamespaces: &addonsv1alpha1.AddonInstallOLMAllNamespaces{
+					AddonInstallOLMCommon: addonsv1alpha1.AddonInstallOLMCommon{
+						Namespace:          "reference-addon",
+						PackageName:        addonName,
+						Channel:            "alpha",
+						CatalogSourceImage: referenceAddonCatalogSourceImageWorking,
+					},
+				},
+			}, addonName),
+			err: nil,
+		},
+	}
+
+	for i, tc := range testCases {
+		tc := tc // pin
+		t.Run(fmt.Sprintf("test case: %d", i), func(t *testing.T) {
+			err := integration.Client.Create(ctx, tc.addon)
+
+			if err == nil {
+				require.NoError(t, err)
+
+				// clean-up addon
+				err = integration.Client.Delete(ctx, tc.addon)
+				require.NoError(t, err)
+
+				err = integration.WaitToBeGone(t, 5*time.Minute, tc.addon)
+				require.NoError(t, err, "wait for Addon to be deleted")
+			} else {
+				assert.EqualValues(t, tc.err, err)
+			}
+		})
+	}
+}
+
+func TestAddonSpecImmutability(t *testing.T) {
+	if !testutil.IsWebhookServerEnabled() {
+		t.Skip("skipping test as webhook server execution is disabled")
+	}
+
+	t.Parallel()
+
+	ctx := context.Background()
+	addonName := "reference-addon-test-install-spec-immutability"
+
+	addon := testutil.NewAddonWithInstallSpec(addonsv1alpha1.AddonInstallSpec{
+		Type: addonsv1alpha1.OLMOwnNamespace,
+		OLMOwnNamespace: &addonsv1alpha1.AddonInstallOLMOwnNamespace{
+			AddonInstallOLMCommon: addonsv1alpha1.AddonInstallOLMCommon{
+				Namespace:          "reference-addon",
+				PackageName:        addonName,
+				Channel:            "alpha",
+				CatalogSourceImage: referenceAddonCatalogSourceImageWorking,
+			},
+		},
+	}, addonName)
+
+	err := integration.Client.Create(ctx, addon)
+	require.NoError(t, err)
+
+	// try to update immutable spec
+	// retry every 10 seconds for 5 minutes
+	err = integration.RetryUntilNoError(time.Minute*5,
```

Comment:
We should be careful to create test functions that will retry any error, because that may hide race conditions in our code.
Also I found this, because I was sure we where not the first to have that issue :) 
https://pkg.go.dev/k8s.io/client-go@v0.22.1/util/retry#RetryOnConflict

```
var DefaultRetry = wait.Backoff{
	Steps:    5,
	Duration: 10 * time.Millisecond,
	Factor:   1.0,
	Jitter:   0.1,
}
```

This is even Jittering a bit and is doing quick but noticable backoff, so we don't DDOS the API Server, while things still stay responsive.